### PR TITLE
Allow basic auth for /oauth/token

### PR DESCRIPTION
**Pull request by stillinbeta**
http://tools.ietf.org/html/rfc6749#section-2.3.1

### CODE REVIEW

File: test/auth_test.rb [hunk 1]
```diff
@@ -209,62 +209,71 @@ def app
       assert_equal 7200, tokens["expires_in"]
     end
 
-    it "accepts an authorization code" do
-      i = 0
-      stub_heroku_api do
-        post("/oauth/tokens") {
-          # only check parameters on the 2nd request, as ID does one OAuth
-          # dance to procure its own token
-          if (i += 1) > 1
```
Comment (stillinbeta) [inline test/auth_test.rb]: I tested this, and the assumption in the comment isn't true. lines 219-221 are never executed with this check in place.

Comment (brandur) [inline test/auth_test.rb]: I think this was probably broken by #91, with the test/comment never fixed.

**Comment by brandur** [comment]
@stillinbeta Thanks! What's this for?

Although in the spec, I've still yet to come across an OAuth 2 client library that depends on this, and absent of a compelling use case, I'm tempted to take spec non-compliance over the additional code complexity.

**Comment by stillinbeta** [comment]
The library I ran into trouble with is the Go one:
https://github.com/golang/oauth2/blob/master/oauth2.go#L279

They do actually have a blacklist of sites that don't support the spec and
they need to do differently, but when I sent a PR they said I needed to
sign some corporate license agreement, and that yak's going to take a long
time to shave.

File: lib/identity/auth.rb [hunk 2]
```diff
@@ -149,6 +149,13 @@ class Auth < Sinatra::Base
         begin
           res = log :create_token, by_proxy: true,
             session_id: @cookie.session_id do
+            # per RFC 6749 section 2.3.1, token endpoint must accept basic auth
+            req = Rack::Auth::Basic::Request.new(request.env)
+            client_secret = if req.provided? && req.basic?
+                              req.credentials.last
```
Comment (brandur) [inline lib/identity/auth.rb]: Can we do something more like:

```
# credentials contain a client ID and a client secret (which is usually included in the request body).
_, client_secret = req.credentials
```

File: lib/identity/auth.rb [hunk 3]
```diff
@@ -149,6 +149,13 @@ class Auth < Sinatra::Base
         begin
           res = log :create_token, by_proxy: true,
             session_id: @cookie.session_id do
+            # per RFC 6749 section 2.3.1, token endpoint must accept basic auth
+            req = Rack::Auth::Basic::Request.new(request.env)
+            client_secret = if req.provided? && req.basic?
```
Comment (brandur) [inline lib/identity/auth.rb]: Can we extract this to a helper method?

**Comment by brandur** [comment]
@stillinbeta Okay, let's do this then.

I left a few minor comments above. Aside from those, can you change `#call_authorize` so that we can support a client ID on the `/oauth/authorize` endpoint as well? My interpretation of the spec is that a consumer is allowed to pass these values by basic authorization anytime it authenticates.

**Comment by stillinbeta** [comment]
@brandur from my reading of the code (and the oauth spec) the client secret isn't involved in the /auth endpoint, so I don't think 2.3.1 applies. See the diagram in https://community.constantcontact.com/t5/Documentation/Authentication-using-OAuth-2-0-Server-and-Client-Flows/ba-p/38313 for details.

My understanding is the /auth endpoint is what produces the login page, so it makes sense that the client secret wouldn't be in play yet. RFCs are hard to read, though, so feel free to correct me!

**Comment by brandur** [comment]
@stillinbeta Ah, that's weird. I assumed that it would be allowed during either phase because clients are allowed to pass their ID in during the token step as dictated by the spec:

```
2.3.1. Client Password


   Clients in possession of a client password MAY use the HTTP Basic
   authentication scheme as defined in [RFC2617] to authenticate with
   the authorization server.  The client identifier is encoded using the
   "application/x-www-form-urlencoded" encoding algorithm per
   Appendix B, and the encoded value is used as the username; the client
   password is encoded using the same algorithm and used as the
   password.  The authorization server MUST support the HTTP Basic
   authentication scheme for authenticating clients that were issued a
   client password.
```

But the server often doesn't need the ID at that point because it's somewhat implied by the grant being exchanged.

Anyway, thanks for refactoring that piece! +1.

**Pull request merged**